src/core/reactor_backend.cc

```diff
@@ -1478,51 +1493,149 @@ class reactor_backend_uring final : public reactor_backend {
         fd.fd.shutdown(how);
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, void* buffer, size_t len) override {
-        return readable(fd).then([this, &fd, buffer, len] () mutable {
-            auto r = fd.fd.read(buffer, len);
-            if (!r) {
-                return read_some(fd, buffer, len);
-            }
-            if (size_t(*r) == len) {
-                fd.speculate_epoll(EPOLLIN);
+        if (fd.take_speculation(POLLIN)) {
+            try {
+                auto r = fd.fd.read(buffer, len);
+                if (r) {
+                    if (size_t(*r) == len) {
+                        fd.speculate_epoll(EPOLLIN);
+                    }
+                    return make_ready_future<size_t>(*r);
+                }
+            } catch (const std::system_error& e) {
+                return make_exception_future<size_t>(e);
             }
-            return make_ready_future<size_t>(*r);
+        }
+        return readable(fd).then([this, &fd, buffer, len] () mutable {
+            class read_completion final : public io_completion {
+                pollable_fd_state& _fd;
+                const size_t _to_read;
+                promise<size_t> _result;
+            public:
+                read_completion(pollable_fd_state& fd, size_t to_read)
+                    : _fd(fd), _to_read(to_read) {}
+                void complete(size_t bytes) noexcept final {
+                    if (bytes == _to_read) {
+                        _fd.speculate_epoll(EPOLLIN);
+                    }
+                    _result.set_value(bytes);
+                    delete this;
+                }
+                void set_exception(std::exception_ptr eptr) noexcept final {
+                    _result.set_exception(eptr);
+                    delete this;
+                }
+                future<size_t> get_future() {
+                    return _result.get_future();
+                }
+            };
+            auto desc = std::make_unique<read_completion>(fd, len);
+            auto req = internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false);
+            return submit_request(std::move(desc), std::move(req));
         });
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, const std::vector<iovec>& iov) override {
-        return readable(fd).then([this, &fd, iov = iov] () mutable {
+        if (fd.take_speculation(POLLIN)) {
             ::msghdr mh = {};
-            mh.msg_iov = &iov[0];
+            mh.msg_iov = const_cast<iovec*>(iov.data());
             mh.msg_iovlen = iov.size();
-            auto r = fd.fd.recvmsg(&mh, 0);
-            if (!r) {
-                return read_some(fd, iov);
+            try {
+                auto r = fd.fd.recvmsg(&mh, MSG_DONTWAIT);
```

Comment:
I think it's confusing that some read_some() overloads use read(), and some read_some() overloads use recvmsg(). Changing an overload can cause a crash (for the inotify use case, if we switch to sendmsg) or reduced performance (if we switch to read).

I think we need separate calls in reactor and reactor_backend for read/write/sendmsg/recvmsg, and then to switch socket calls to read/write.

Ah, but I see the problem already exists in master, so I won't require it as a prerequisite.

Reply:
it is. i had a hard time getting my head around this. will fix this in a follow up change.